src/pczt/verify.rs

```diff
@@ -0,0 +1,176 @@
+use crate::{
+    keys::{FullViewingKey, SpendValidatingKey},
+    note::{ExtractedNoteCommitment, Rho},
+    value::ValueCommitment,
+    Note,
+};
+
+impl super::Action {
+    /// Verifies that the `cv_net` field is consistent with the note fields.
+    ///
+    /// Requires that the following optional fields are set:
+    /// - `spend.value`
+    /// - `output.value`
+    /// - `rcv`
+    pub fn verify_cv_net(&self) -> Result<(), VerifyError> {
+        let spend_value = self.spend().value.ok_or(VerifyError::MissingValue)?;
+        let output_value = self.output().value.ok_or(VerifyError::MissingValue)?;
+        let rcv = self
+            .rcv
+            .clone()
+            .ok_or(VerifyError::MissingValueCommitTrapdoor)?;
+
+        let cv_net = ValueCommitment::derive(spend_value - output_value, rcv);
+        if cv_net.to_bytes() == self.cv_net.to_bytes() {
+            Ok(())
+        } else {
+            Err(VerifyError::InvalidValueCommitment)
+        }
+    }
+}
+
+impl super::Spend {
+    /// Returns the [`FullViewingKey`] to use when validating this note.
+    ///
+    /// Handles dummy notes when the `value` field is set.
+    fn fvk_for_validation<'a>(
+        &'a self,
+        expected_fvk: Option<&'a FullViewingKey>,
+    ) -> Result<&'a FullViewingKey, VerifyError> {
+        match (expected_fvk, self.fvk.as_ref(), self.value.as_ref()) {
+            // Dummy notes use random FVKs, which must be provided.
+            (_, Some(fvk), Some(value)) if value.inner() == 0 => Ok(fvk),
+            (_, None, Some(value)) if value.inner() == 0 => Err(VerifyError::MissingFullViewingKey),
+            // If the FVK field has been pruned, assume the caller provided the correct FVK.
+            (Some(expected_fvk), None, _) => Ok(expected_fvk),
+            // This is not a dummy note; if the FVK field is present, it must match.
+            (Some(expected_fvk), Some(fvk), _) if fvk == expected_fvk => Ok(fvk),
+            (Some(_), Some(_), _) => Err(VerifyError::MismatchedFullViewingKey),
+            (None, Some(fvk), _) => Ok(fvk),
+            (None, None, _) => Err(VerifyError::MissingFullViewingKey),
+        }
```

Comment:
What is the motivation for treating dummy notes differently? The main effect seems to be that if we pass a different `expected_fvk` than `self.fvk` when both are `Some`, when we will allow that for a dummy note. But why?

If not for that case, we might as well do the simpler:
```suggestion
        match (expected_fvk, self.fvk.as_ref()) {
            (Some(expected_fvk), Some(fvk)) if fvk == expected_fvk => Ok(fvk),
            (Some(_), Some(_)) => Err(VerifyError::MismatchedFullViewingKey),
            (Some(expected_fvk), None) => Ok(expected_fvk),
            (None, Some(fvk)) => Ok(fvk),
            (None, None) => Err(VerifyError::MissingFullViewingKey),
        }
```